internal/kafka/test/integration/kafka_expiration_test.go

```diff
@@ -0,0 +1,513 @@
+			mockkafka.With(mockkafka.OWNER, dummyOrgUser),
+			mockkafka.With(mockkafka.STATUS, constants.KafkaRequestStatusReady.String()),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_ID, "dummy-sa-client-id"),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_SECRET, "dummy-sa-client-secret"),
+			mockkafka.With(mockkafka.INSTANCE_TYPE, types.STANDARD.String()),
+			mockkafka.With(mockkafka.ACTUAL_KAFKA_BILLING_MODEL, bmExpiredWithoutGracePeriodID),
+			mockkafka.With(mockkafka.DESIRED_KAFKA_BILLING_MODEL, bmExpiredWithoutGracePeriodID),
+			mockkafka.With(mockkafka.CLUSTER_ID, testCluster.ClusterID),
+			// simulate kafka with a null expires_at value
+			mockkafka.WithExpiresAt(sql.NullTime{}),
+		),
+		// should expire and deleted straight away as the given quota is expired and has no grace period allowance
+		mockkafka.BuildKafkaRequest(
+			mockkafka.WithPredefinedTestValues(),
+			mockkafka.With(mockkafka.NAME, "kafka-to-expire-2"),
+			mockkafka.With(mockkafka.ORGANISATION_ID, dummyOrgID),
+			mockkafka.With(mockkafka.OWNER, dummyOrgUser),
+			mockkafka.With(mockkafka.STATUS, constants.KafkaRequestStatusReady.String()),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_ID, "dummy-sa-client-id"),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_SECRET, "dummy-sa-client-secret"),
+			mockkafka.With(mockkafka.INSTANCE_TYPE, types.STANDARD.String()),
+			mockkafka.With(mockkafka.ACTUAL_KAFKA_BILLING_MODEL, bmExpiredWithoutGracePeriodID),
+			mockkafka.With(mockkafka.DESIRED_KAFKA_BILLING_MODEL, bmExpiredWithoutGracePeriodID),
+			mockkafka.With(mockkafka.CLUSTER_ID, testCluster.ClusterID),
+			// simulate kafka with a zero time.Time expires_at value
+			mockkafka.WithExpiresAt(sql.NullTime{Time: time.Time{}, Valid: true}),
+		),
+		// should expire as quota for given billing model is no longer entitled to the user
+		mockkafka.BuildKafkaRequest(
+			mockkafka.WithPredefinedTestValues(),
+			mockkafka.With(mockkafka.NAME, "kafka-to-expire-3"),
+			mockkafka.With(mockkafka.ORGANISATION_ID, dummyOrgID),
+			mockkafka.With(mockkafka.OWNER, dummyOrgUser),
+			mockkafka.With(mockkafka.STATUS, constants.KafkaRequestStatusReady.String()),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_ID, "dummy-sa-client-id"),
+			mockkafka.With(mockkafka.CANARY_SERVICE_ACCOUNT_CLIENT_SECRET, "dummy-sa-client-secret"),
+			mockkafka.With(mockkafka.INSTANCE_TYPE, types.STANDARD.String()),
+			mockkafka.With(mockkafka.ACTUAL_KAFKA_BILLING_MODEL, bmNoLongerEntitledID),
+			mockkafka.With(mockkafka.DESIRED_KAFKA_BILLING_MODEL, bmNoLongerEntitledID),
+			mockkafka.With(mockkafka.CLUSTER_ID, testCluster.ClusterID),
+			mockkafka.WithExpiresAt(sql.NullTime{}),
+		),
+	}
+
+	// create dummy kafkas
+	var kafkas []*dbapi.KafkaRequest
+	kafkas = append(kafkas, kafkasToRemain...)
+	kafkas = append(kafkas, kafkasToExpire...)
+	if err := db.Create(&kafkas).Error; err != nil {
+		g.Expect(err).NotTo(gomega.HaveOccurred())
```

Comment:
This will always fail

Reply:
good catch!

Reply:
I guess this might be actually a strange way to intentionally make it fail because of the `return` at the end.
The way it should be rewritten is:
``
err := db.Create(&kafkas).Error;
g.Expect(err).NotTo(gomega.HaveOccurred())
``

Reply:
updated there, thanks!